File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/ContentTypeAttributeInfo.kt
package com.jetbrains.signatureverifier.serialization

import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.cms.Attribute

// 1.2.840.113549.1.9.3
@Serializable
data class ContentTypeAttributeInfo(
  val identifier: TextualInfo,
  val content: List<TextualInfo>
) : AttributeInfo {
  override fun toAttributeDLSequence(): DLSequence =
    listOf(
      identifier.toPrimitive(),
      content.map { it.toPrimitive() }.toDLSet()
    ).toDLSequence()


  constructor(attribute: Attribute) : this(
    TextualInfo.getInstance(attribute.attrType),
    attribute.attributeValues.map { TextualInfo.getInstance(it) }
  )
}

File: net/JetBrains.Serialization/src/Dataholders/ContentTypeAttributeInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Attribute = Org.BouncyCastle.Asn1.Cms.Attribute;
using JetBrains.Serialization;

[JsonObject(MemberSerialization.Fields)]
public class ContentTypeAttributeInfo : AttributeInfo
{
    public override TextualInfo Identifier { get; }
    public List<TextualInfo> Content { get; }

    public ContentTypeAttributeInfo(Attribute attribute)
    {
        Identifier = TextualInfo.GetInstance(attribute.AttrType);
        Content = attribute.AttrValues.ToArray().Select(item => TextualInfo.GetInstance(item)).ToList();
    }

    public override Asn1Encodable GetPrimitiveContent() => Content.ToPrimitiveDerSet();
}